File: src/drivers/lapic.cpp
#include <drivers/lapic.hpp>
#include <drivers/pic.hpp>
#include <cpu/cpu.hpp>
#include <cpuid.h>
#include <lib/logger.hpp>
#include <drivers/madt.hpp>
#include <cpu/idt.hpp>
#include <mem/mapper.hpp>
#include <mem/pmm.hpp>
#include <lib/interrupts.hpp>
#include <drivers/pit.hpp>
#include <lib/timer.hpp>
#include <cpu/smp.hpp>
#include <uacpi/kernel_api.h>
#include <assert.h>
using namespace lapic;

constexpr int TDIV1 = 0xB;
constexpr int TDIV2 = 0;
constexpr int TDIV4 = 1;
constexpr int TDIV8 = 2;
constexpr int TDIV16 = 3;
constexpr int TDIV32 = 8;
constexpr int TDIV64 = 9;
constexpr int TDIV128 = 0xA;
constexpr int TDIV = TDIV4;
constexpr int IA32_APIC_BASE = 0x1B;

constexpr int SPURIOUS_OFF = 0xF0;
constexpr int ICRLO_OFF = 0x300;
constexpr int ICRHI_OFF = 0x310;
constexpr int INITIAL_COUNT_OFF = 0x380;
constexpr int CUR_COUNT_OFF = 0x390;
constexpr int CONFIG_OFF = 0x3E0;
constexpr int LVT_OFFSET = 0x320;

bool x2mode, tsc;
static uintptr_t addr;
std::size_t lapic::ms_ticks;
std::size_t lapic::ticks;
bool lapic::initialized = false;
uint8_t spurious_vec;

void calibrate();

void lapic::bsp_init() {
    uint64_t msr = cpu::rdmsr(IA32_APIC_BASE);
    assert(msr & (1 << 11));
    
    uint32_t ignored, ecx = 0, edx = 0;
    __cpuid(1, ignored, ignored, ecx, edx);

    if (ecx & (1 << 21))
        x2mode = true;
    else if (edx & (1 << 9)) {
        x2mode = false;
        addr = virt_addr(msr & 0xffffffffff000);
    } else return;

    spurious_vec = idt::set_handler([](cpu::status *status) { return status; });
    timer::irq = intr::IRQ(idt::set_handler([](cpu::status *status) {
        if (smp::this_cpu()->id == smp::bsp_id)
            lapic::ticks++;
        intr::eoi(0);
        return status;
    }));

    write_reg(SPURIOUS_OFF, (1 << 8) | spurious_vec);
    logger::info("LAPIC", "Initialized %sAPIC", x2mode ? "x2" : "x");
    calibrate();
    initialized = true;
}

void lapic::ap_init() {
    write_reg(SPURIOUS_OFF, (1 << 8) | spurious_vec);
}

void calibrate() {
    write_reg(INITIAL_COUNT_OFF, 0);
    write_reg(CONFIG_OFF, TDIV);

    write_reg(INITIAL_COUNT_OFF, (uint32_t) - 1);
    timer::sleep(100);
    uint32_t cur_ticks = read_reg(CUR_COUNT_OFF);

    uint32_t time_elapsed = ((uint32_t) - 1) - cur_ticks;
    ms_ticks = time_elapsed / 100;
    logger::verbose("LAPIC", "APIC ticks per ms: %u", ms_ticks);
}

void lapic::start(std::size_t rate) {
    pit::stop();
    write_reg(LVT_OFFSET, intr::VEC(timer::irq) | (Periodic << 17));
    write_reg(INITIAL_COUNT_OFF, rate);
    write_reg(CONFIG_OFF, TDIV);
}

void lapic::stop() {
    write_reg(INITIAL_COUNT_OFF, 0);
}

void lapic::ipi(uint32_t flags, uint32_t id) {
    if (x2mode)
        write_reg(ICRLO_OFF, (static_cast<uint64_t>(id) << 32) | flags | (1 << 14));
    else {
        write_reg(ICRHI_OFF, id << 24);
        write_reg(ICRLO_OFF, flags | (1 << 14));
    }
}

uint64_t lapic::read_reg(uint32_t off) {
    if (x2mode)
        return cpu::rdmsr((off >> 4) + 0x800);
    else
        return *(volatile uint32_t*) (addr + off);
}

void lapic::write_reg(uint32_t off, uint64_t val) {
    if (x2mode)
        cpu::wrmsr((off >> 4) + 0x800, val);
    else
        *(volatile uint32_t*)(addr + off) = val;
}

File: src/drivers/lapic.zig
const kernel = @import("root");
const log = @import("std").log.scoped(.lapic);
const cpu = kernel.drivers.cpu;

const MSR = 0x1B;
const SPURIOUS_VEC = 0xFF;

const TPR_OFF = 0x80;
const EOI_OFF = 0xB0;
const SPURIOUS_OFF = 0xF0;
const CONFIG_OFF = 0x3E0;
const ICRLO = 0x300;
const ICRHI = 0x310;

// 0b0 - Divide by 2
// 0b1 - Divide by 4
// 0b10 - Divide by 8
// 0b11 - Divide by 16
// 0b1000 - Divide by 32
// 0b1001 - Divide by 64
// 0b1010 - Divide by 128
// 0b1011 - Divide by 1
const TDIV = 1;

var addr: usize = undefined;
pub var read_reg: *const fn (off: u32) u64 = undefined;
pub var write_reg: *const fn (off: u32, val: u64) void = undefined;

pub var Task = kernel.Task{
    .name = "Local APIC",
    .init = bsp_init,
    .dependencies = &.{
        .{ .task = &kernel.drivers.idt.Task },
        .{ .task = &kernel.lib.mem.KMapperTask },
    },
};

pub var TaskAP = kernel.Task{
    .name = "Local APIC (AP)",
    .init = ap_init,
    .dependencies = &.{
        .{ .task = &kernel.lib.mem.KMapperTaskAP },
    },
};

fn bsp_init() kernel.Task.Ret {
    var msr = cpu.rdmsr(MSR) | (1 << 11);
    const cpuid = cpu.cpuid(1, 0);
    if (cpuid.ecx & (1 << 21) > 0) {
        msr |= 1 << 10;
        read_reg = x2apic_read_reg;
        write_reg = x2apic_write_reg;
        log.debug("Found x2APIC", .{});
    } else if (cpuid.edx & (1 << 9) > 0) {
        addr = msr & ~@as(u64, 0xFFF);
        kernel.lib.mem.kmapper.map(addr, addr, (1 << 63) | 0b11);
        read_reg = xapic_read_reg;
        write_reg = xapic_write_reg;
        log.debug("Found xAPIC", .{});
    } else return .failed;
    cpu.wrmsr(MSR, msr);

    kernel.drivers.idt.vec2handler(SPURIOUS_VEC).reserved = true;
    write_reg(SPURIOUS_OFF, (@as(u32, 1) << 8) | SPURIOUS_VEC);
    write_reg(TPR_OFF, 0);
    write_reg(CONFIG_OFF, TDIV);
    return .success;
}

fn ap_init() kernel.Task.Ret {
    var msr = cpu.rdmsr(MSR) | (1 << 11);
    if (read_reg == x2apic_read_reg) msr |= 1 << 10;
    cpu.wrmsr(MSR, msr);

    write_reg(SPURIOUS_OFF, (@as(u32, 1) << 8) | SPURIOUS_VEC);
    write_reg(TPR_OFF, 0);
    write_reg(CONFIG_OFF, TDIV);
    return .success;
}

pub fn ipi(dest: u8, vec: u8) void {
    if (read_reg == x2apic_read_reg) {
        write_reg(ICRLO, (@as(u64, dest) << 56) | vec);
    } else {
        write_reg(ICRHI, @as(u32, dest) << 24);
        write_reg(ICRLO, vec);
    }
}

pub inline fn eoi() void {
    write_reg(EOI_OFF, 0);
}

fn xapic_write_reg(off: u32, val: u64) void {
    @as(*u32, @ptrFromInt(addr + off)).* = @truncate(val);
}

fn x2apic_write_reg(off: u32, val: u64) void {
    cpu.wrmsr((off >> 4) + 0x800, val);
}

fn xapic_read_reg(off: u32) u64 {
    return @intCast(@as(*const u32, @ptrFromInt(addr + off)).*);
}

fn x2apic_read_reg(off: u32) u64 {
    return cpu.rdmsr((off >> 4) + 0x800);
}
